crates/conjure-cp-core/src/ast/domains.rs

```diff
@@ -51,6 +51,16 @@ impl<A: Ord> Range<A> {
             Range::UnboundedL(_) => None,
         }
     }
+
+    /// Returns the upper bound of the range, if it has one
+    pub fn upper_bound(&self) -> Option<&A> {
```

Comment:
remark: maybe merge lower_bound and upper_bound functions like this?
```suggestion
        /// Returns both bounds of the range (lower, upper).
    pub fn bounds(&self) -> (Option<&A>, Option<&A>) {
        match self {
            Range::Single(a) => (Some(a), Some(a)),
            Range::Bounded(a, b) => (Some(a), Some(b)),
            Range::UnboundedR(a) => (Some(a), None),
            Range::UnboundedL(a) => (None, Some(a)),
        }
    }

    /// Returns the lower bound of the range, if it has one
    pub fn lower_bound(&self) -> Option<&A> {
        self.bounds().0
    }

    /// Returns the upper bound of the range, if it has one
    pub fn upper_bound(&self) -> Option<&A> {
        self.bounds().1
    }
```